File: apps/web/tests/test-runner.js
#!/usr/bin/env node

/**
 * Comprehensive test runner for frontend file upload functionality
 */

const { execSync, spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

class FrontendTestRunner {
  constructor() {
    this.projectRoot = path.resolve(__dirname, '../../../');
    this.webRoot = path.resolve(this.projectRoot, 'apps/web');
    this.testResults = {
      timestamp: new Date().toISOString(),
      suites: {},
      coverage: {},
      summary: { passed: 0, failed: 0, skipped: 0 }
    };
  }

  async runUnitTests(verbose = false) {
    console.log('🧪 Running Frontend Unit Tests...');
    
    try {
      const cmd = [
        'npm', 'test',
        '--',
        '--coverage',
        '--coverageReporters=json,html,text-summary',
        '--coverageDirectory=coverage',
        '--testMatch=**/*.test.tsx',
        '--testMatch=**/*.test.ts',
        verbose ? '--verbose' : '--silent'
      ];

      const result = await this.runCommand(cmd, this.webRoot);
      
      if (result.success) {
        console.log('✅ Frontend unit tests passed');
        this.testResults.suites.unit = {
          status: 'passed',
          output: result.stdout
        };
        return true;
      } else {
        console.log('❌ Frontend unit tests failed');
        console.log(result.stdout);
        console.log(result.stderr);
        this.testResults.suites.unit = {
          status: 'failed',
          output: result.stdout,
          error: result.stderr
        };
        return false;
      }
    } catch (error) {
      console.log(`❌ Error running unit tests: ${error.message}`);
      this.testResults.suites.unit = {
        status: 'error',
        error: error.message
      };
      return false;
    }
  }

  async runComponentTests(verbose = false) {
    console.log('🎨 Running Component Tests...');
    
    try {
      const cmd = [
        'npm', 'test',
        '--',
        '--testMatch=**/components/**/*.test.tsx',
        '--coverage',
        '--coverageDirectory=coverage/components',
        verbose ? '--verbose' : '--silent'
      ];

      const result = await this.runCommand(cmd, this.webRoot);
      
      if (result.success) {
        console.log('✅ Component tests passed');
        this.testResults.suites.components = {
          status: 'passed',
          output: result.stdout
        };
        return true;
      } else {
        console.log('❌ Component tests failed');
        console.log(result.stdout);
        this.testResults.suites.components = {
          status: 'failed',
          output: result.stdout,
          error: result.stderr
        };
        return false;
      }
    } catch (error) {
      console.log(`❌ Error running component tests: ${error.message}`);
      this.testResults.suites.components = {
        status: 'error',
        error: error.message
      };
      return false;
    }
  }

  async runE2ETests(verbose = false) {
    console.log('🎭 Running E2E Tests...');
    
    try {
      const cmd = [
        'npx', 'playwright', 'test',
        'tests/e2e/file-upload-workflow.spec.ts',
        '--reporter=json',
        '--output-dir=test-results',
        verbose ? '' : '--quiet'
      ].filter(Boolean);

      const result = await this.runCommand(cmd, this.webRoot);
      
      if (result.success) {
        console.log('✅ E2E tests passed');
        this.testResults.suites.e2e = {
          status: 'passed',
          output: result.stdout
        };
        return true;
      } else {
        console.log('❌ E2E tests failed');
        console.log(result.stdout);
        this.testResults.suites.e2e = {
          status: 'failed',
          output: result.stdout,
          error: result.stderr
        };
        return false;
      }
    } catch (error) {
      console.log(`❌ Error running E2E tests: ${error.message}`);
      this.testResults.suites.e2e = {
        status: 'error',
        error: error.message
      };
      return false;
    }
  }

  async runAccessibilityTests(verbose = false) {
    console.log('♿ Running Accessibility Tests...');
    
    try {
      const cmd = [
        'npm', 'test',
        '--',
        '--testMatch=**/*.a11y.test.tsx',
        '--testMatch=**/*.accessibility.test.tsx',
        verbose ? '--verbose' : '--silent'
      ];

      const result = await this.runCommand(cmd, this.webRoot);
      
      if (result.success) {
        console.log('✅ Accessibility tests passed');
        this.testResults.suites.accessibility = {
          status: 'passed',
          output: result.stdout
        };
        return true;
      } else {
        console.log('⚠️  Some accessibility tests failed');
        console.log(result.stdout);
        this.testResults.suites.accessibility = {
          status: 'warning',
          output: result.stdout,
          error: result.stderr
        };
        return true; // Don't fail CI for accessibility issues
      }
    } catch (error) {
      console.log(`⚠️  Error running accessibility tests: ${error.message}`);
      this.testResults.suites.accessibility = {
        status: 'error',
        error: error.message
      };
      return true; // Don't fail CI for accessibility issues
    }
  }

  async runPerformanceTests(verbose = false) {
    console.log('⚡ Running Performance Tests...');
    
    try {
      // Create a simple performance test
      const perfTestContent = `
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { FileUpload } from '@/components/dashboard/upload/FileUpload';

// Performance benchmark test
test('FileUpload component renders within performance budget', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: { retry: false },
      mutations: { retry: false },
    },
  });

  const start = performance.now();
  
  render(
    <QueryClientProvider client={queryClient}>
      <FileUpload onUploadSuccess={jest.fn()} />
    </QueryClientProvider>
  );
  
  const end = performance.now();
  const renderTime = end - start;
  
  // Component should render within 100ms
  expect(renderTime).toBeLessThan(100);
  console.log(\`FileUpload rendered in \${renderTime.toFixed(2)}ms\`);
});

test('Multiple file selections should not cause memory leaks', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: { retry: false },
      mutations: { retry: false },
    },
  });

  const initialMemory = (performance as any).memory?.usedJSHeapSize || 0;
  
  // Render multiple times to simulate file selections
  for (let i = 0; i < 10; i++) {
    const { unmount } = render(
      <QueryClientProvider client={queryClient}>
        <FileUpload onUploadSuccess={jest.fn()} />
      </QueryClientProvider>
    );
    unmount();
  }
  
  const finalMemory = (performance as any).memory?.usedJSHeapSize || 0;
  const memoryIncrease = finalMemory - initialMemory;
  
  // Memory increase should be minimal (less than 5MB)
  expect(memoryIncrease).toBeLessThan(5 * 1024 * 1024);
});
`;

      // Write performance test
      const perfTestPath = path.join(this.webRoot, 'tests/performance/fileupload.performance.test.tsx');
      const perfTestDir = path.dirname(perfTestPath);
      
      if (!fs.existsSync(perfTestDir)) {
        fs.mkdirSync(perfTestDir, { recursive: true });
      }
      
      fs.writeFileSync(perfTestPath, perfTestContent);

      const cmd = [
        'npm', 'test',
        '--',
        '--testMatch=**/performance/*.test.tsx',
        verbose ? '--verbose' : '--silent'
      ];

      const result = await this.runCommand(cmd, this.webRoot);
      
      if (result.success) {
        console.log('✅ Performance tests passed');
        this.testResults.suites.performance = {
          status: 'passed',
          output: result.stdout
        };
        return true;
      } else {
        console.log('⚠️  Some performance benchmarks may have failed');
        console.log(result.stdout);
        this.testResults.suites.performance = {
          status: 'warning',
          output: result.stdout,
          error: result.stderr
        };
        return true; // Don't fail CI for performance issues
      }
    } catch (error) {
      console.log(`⚠️  Error running performance tests: ${error.message}`);
      this.testResults.suites.performance = {
        status: 'error',
        error: error.message
      };
      return true; // Don't fail CI for performance issues
    }
  }

  async generateCoverageReport() {
    console.log('📊 Generating Coverage Report...');
    
    try {
      const coveragePath = path.join(this.webRoot, 'coverage/coverage-summary.json');
      
      if (fs.existsSync(coveragePath)) {
        const coverageData = JSON.parse(fs.readFileSync(coveragePath, 'utf8'));
        
        if (coverageData.total && coverageData.total.statements) {
          const coverage = coverageData.total.statements.pct;
          this.testResults.coverage.statements = coverage;
          
          if (coverage >= 90) {
            console.log(`✅ Excellent coverage: ${coverage}%`);
          } else if (coverage >= 80) {
            console.log(`✅ Good coverage: ${coverage}%`);
          } else if (coverage >= 70) {
            console.log(`⚠️  Moderate coverage: ${coverage}%`);
          } else {
            console.log(`❌ Low coverage: ${coverage}%`);
          }
        }
      }
    } catch (error) {
      console.log(`⚠️  Error generating coverage report: ${error.message}`);
    }
  }

  generateFinalReport() {
    console.log('\n' + '='.repeat(60));
    console.log('📋 FRONTEND TEST REPORT');
    console.log('='.repeat(60));
    
    // Count results
    const passed = Object.values(this.testResults.suites).filter(suite => suite.status === 'passed').length;
    const failed = Object.values(this.testResults.suites).filter(suite => suite.status === 'failed').length;
    const warnings = Object.values(this.testResults.suites).filter(suite => suite.status === 'warning').length;
    const errors = Object.values(this.testResults.suites).filter(suite => suite.status === 'error').length;
    
    this.testResults.summary = {
      passed,
      failed,
      warnings,
      errors,
      total: Object.keys(this.testResults.suites).length
    };
    
    // Print summary
    Object.entries(this.testResults.suites).forEach(([suiteName, result]) => {
      const statusIcons = {
        passed: '✅',
        failed: '❌',
        warning: '⚠️ ',
        error: '💥'
      };
      
      const icon = statusIcons[result.status] || '❓';
      console.log(`${icon} ${suiteName.charAt(0).toUpperCase() + suiteName.slice(1)} Tests: ${result.status.toUpperCase()}`);
    });
    
    if (this.testResults.coverage.statements) {
      console.log(`📊 Test Coverage: ${this.testResults.coverage.statements}%`);
    }
    
    console.log(`\n🎯 Summary: ${passed} passed, ${failed} failed, ${warnings} warnings, ${errors} errors`);
    
    // Save detailed report
    const reportPath = path.join(this.webRoot, 'test-report.json');
    fs.writeFileSync(reportPath, JSON.stringify(this.testResults, null, 2));
    console.log(`📁 Detailed report saved to: ${reportPath}`);
    
    if (failed > 0) {
      console.log('\n❌ Some tests failed. Check the output above for details.');
      return false;
    } else if (errors > 0) {
      console.log('\n⚠️  Some test suites had errors. Check the output above.');
      return false;
    } else {
      console.log('\n✅ All tests passed successfully!');
      return true;
    }
  }

  runCommand(cmd, cwd) {
    return new Promise((resolve) => {
      const child = spawn(cmd[0], cmd.slice(1), {
        cwd,
        stdio: ['pipe', 'pipe', 'pipe']
      });

      let stdout = '';
      let stderr = '';

      child.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      child.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      child.on('close', (code) => {
        resolve({
          success: code === 0,
          stdout,
          stderr,
          code
        });
      });
    });
  }
}

async function main() {
  const args = process.argv.slice(2);
  const verbose = args.includes('--verbose') || args.includes('-v');
  const unitOnly = args.includes('--unit-only');
  const componentOnly = args.includes('--component-only');
  const e2eOnly = args.includes('--e2e-only');
  const accessibilityOnly = args.includes('--accessibility-only');
  const performanceOnly = args.includes('--performance-only');
  const skipCoverage = args.includes('--skip-coverage');
  
  const runner = new FrontendTestRunner();
  
  console.log('🚀 Starting Frontend Test Suite for File Upload Functionality');
  console.log(`📁 Web Root: ${runner.webRoot}`);
  console.log(`⏰ Started at: ${new Date().toISOString()}`);
  console.log();
  
  let success = true;
  
  // Run selected test suites
  if (unitOnly) {
    success = await runner.runUnitTests(verbose);
  } else if (componentOnly) {
    success = await runner.runComponentTests(verbose);
  } else if (e2eOnly) {
    success = await runner.runE2ETests(verbose);
  } else if (accessibilityOnly) {
    success = await runner.runAccessibilityTests(verbose);
  } else if (performanceOnly) {
    success = await runner.runPerformanceTests(verbose);
  } else {
    // Run all test suites
    success &= await runner.runUnitTests(verbose);
    success &= await runner.runComponentTests(verbose);
    success &= await runner.runE2ETests(verbose);
    success &= await runner.runAccessibilityTests(verbose);
    success &= await runner.runPerformanceTests(verbose);
  }
  
  // Generate coverage report
  if (!skipCoverage) {
    await runner.generateCoverageReport();
  }
  
  // Generate final report
  const finalSuccess = runner.generateFinalReport();
  
  // Exit with appropriate code
  process.exit((success && finalSuccess) ? 0 : 1);
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { FrontendTestRunner };

File: apps/api/tests/test_runner.py
#!/usr/bin/env python3
"""
Comprehensive test runner for file upload functionality
"""

import os
import sys
import subprocess
import argparse
from pathlib import Path
from typing import List, Dict, Any
import json
from datetime import datetime


class TestRunner:
    """Enhanced test runner with coverage and reporting"""
    
    def __init__(self, project_root: str):
        self.project_root = Path(project_root)
        self.api_root = self.project_root / "apps" / "api"
        self.test_results: Dict[str, Any] = {
            "timestamp": datetime.now().isoformat(),
            "suites": {},
            "coverage": {},
            "summary": {"passed": 0, "failed": 0, "skipped": 0}
        }
    
    def run_unit_tests(self, verbose: bool = False) -> bool:
        """Run all unit tests with coverage"""
        print("🧪 Running Unit Tests...")
        
        cmd = [
            "python", "-m", "pytest",
            "tests/unit/",
            "--cov=src",
            "--cov-report=json:coverage_unit.json",
            "--cov-report=html:htmlcov_unit",
            "--cov-report=term-missing",
            "--json-report", 
            "--json-report-file=test_results_unit.json",
            "-v" if verbose else "-q"
        ]
        
        try:
            result = subprocess.run(
                cmd,
                cwd=self.api_root,
                capture_output=True,
                text=True,
                env={**os.environ, **{
                    "SECRET_KEY": "test_secret_key_that_is_32_chars_long",
                    "JWT_SECRET_KEY": "test_jwt_secret_key_that_is_32_chars_long",
                    "DATABASE_URL": "sqlite:///test.db",
                    "DATABASE_URL_ASYNC": "sqlite+aiosqlite:///test.db",
                    "OPENAI_API_KEY": "test_key"
                }}
            )
            
            if result.returncode == 0:
                print("✅ Unit tests passed")
                self.test_results["suites"]["unit"] = {
                    "status": "passed",
                    "output": result.stdout
                }
                return True
            else:
                print("❌ Unit tests failed")
                print(result.stdout)
                print(result.stderr)
                self.test_results["suites"]["unit"] = {
                    "status": "failed",
                    "output": result.stdout,
                    "error": result.stderr
                }
                return False
                
        except Exception as e:
            print(f"❌ Error running unit tests: {e}")
            self.test_results["suites"]["unit"] = {
                "status": "error",
                "error": str(e)
            }
            return False
    
    def run_integration_tests(self, verbose: bool = False) -> bool:
        """Run integration tests"""
        print("🔗 Running Integration Tests...")
        
        cmd = [
            "python", "-m", "pytest",
            "tests/integration/",
            "--cov=src",
            "--cov-report=json:coverage_integration.json",
            "--cov-report=html:htmlcov_integration",
            "--json-report",
            "--json-report-file=test_results_integration.json",
            "-v" if verbose else "-q"
        ]
        
        try:
            result = subprocess.run(
                cmd,
                cwd=self.api_root,
                capture_output=True,
                text=True,
                env={**os.environ, **{
                    "SECRET_KEY": "GE3j2VqzbyfloO6oqa4zJqmQcJuMgKdryNUGdnJl0p4",
                    "JWT_SECRET_KEY": "GE3j2VqzbyfloO6oqa4zJqmQcJuMgKdryNUGdnJl0p4",
                    "DATABASE_URL": "postgresql://postgres:password@localhost:5433/xm_port_dev",
                    "DATABASE_URL_ASYNC": "postgresql+asyncpg://postgres:password@localhost:5433/xm_port_dev",
                    "OPENAI_API_KEY": "dummy_key_for_test"
                }}
            )
            
            if result.returncode == 0:
                print("✅ Integration tests passed")
                self.test_results["suites"]["integration"] = {
                    "status": "passed",
                    "output": result.stdout
                }
                return True
            else:
                print("❌ Integration tests failed")
                print(result.stdout)
                print(result.stderr)
                self.test_results["suites"]["integration"] = {
                    "status": "failed",
                    "output": result.stdout,
                    "error": result.stderr
                }
                return False
                
        except Exception as e:
            print(f"❌ Error running integration tests: {e}")
            self.test_results["suites"]["integration"] = {
                "status": "error",
                "error": str(e)
            }
            return False
    
    def run_security_tests(self, verbose: bool = False) -> bool:
        """Run security-focused tests"""
        print("🔒 Running Security Tests...")
        
        cmd = [
            "python", "-m", "pytest",
            "tests/unit/test_file_upload_security.py",
            "--json-report",
            "--json-report-file=test_results_security.json",
            "-v" if verbose else "-q"
        ]
        
        try:
            result = subprocess.run(
                cmd,
                cwd=self.api_root,
                capture_output=True,
                text=True,
                env={**os.environ, **{
                    "SECRET_KEY": "test_secret_key_that_is_32_chars_long",
                    "JWT_SECRET_KEY": "test_jwt_secret_key_that_is_32_chars_long",
                    "DATABASE_URL": "sqlite:///test.db",
                    "DATABASE_URL_ASYNC": "sqlite+aiosqlite:///test.db",
                    "OPENAI_API_KEY": "test_key"
                }}
            )
            
            if result.returncode == 0:
                print("✅ Security tests passed")
                self.test_results["suites"]["security"] = {
                    "status": "passed",
                    "output": result.stdout
                }
                return True
            else:
                print("❌ Security tests failed")
                print(result.stdout)
                print(result.stderr)
                self.test_results["suites"]["security"] = {
                    "status": "failed",
                    "output": result.stdout,
                    "error": result.stderr
                }
                return False
                
        except Exception as e:
            print(f"❌ Error running security tests: {e}")
            self.test_results["suites"]["security"] = {
                "status": "error",
                "error": str(e)
            }
            return False
    
    def run_performance_tests(self, verbose: bool = False) -> bool:
        """Run performance benchmarks"""
        print("⚡ Running Performance Tests...")
        
        # Create a simple performance test
        perf_test_content = '''
import pytest
import time
import asyncio
from src.services.file_processing import FileProcessingService
from unittest.mock import Mock

@pytest.mark.benchmark
async def test_file_validation_performance():
    """Benchmark file validation performance"""
    mock_db = Mock()
    service = FileProcessingService(mock_db)
    
    # Create test content
    csv_content = "Product Description,Quantity,Unit,Value,Origin Country,Unit Price\\n"
    csv_content += "\\n".join([f'"Product {i}",{i},pieces,{i*5.0},USA,5.00' for i in range(1000)])
    
    start_time = time.time()
    # Performance test would go here
    end_time = time.time()
    
    processing_time = end_time - start_time
    assert processing_time < 5.0, f"Validation took {processing_time}s, should be < 5s"

def test_memory_usage():
    """Test memory usage stays within bounds"""
    import psutil
    import os
    
    process = psutil.Process(os.getpid())
    initial_memory = process.memory_info().rss / 1024 / 1024  # MB
    
    # Simulate large file processing
    large_data = "x" * (10 * 1024 * 1024)  # 10MB
    
    current_memory = process.memory_info().rss / 1024 / 1024  # MB
    memory_increase = current_memory - initial_memory
    
    assert memory_increase < 50, f"Memory increased by {memory_increase}MB, should be < 50MB"
'''
        
        # Write performance test file
        perf_test_file = self.api_root / "tests" / "performance" / "test_performance.py"
        perf_test_file.parent.mkdir(exist_ok=True)
        
        with open(perf_test_file, 'w') as f:
            f.write(perf_test_content)
        
        cmd = [
            "python", "-m", "pytest",
            "tests/performance/",
            "--json-report",
            "--json-report-file=test_results_performance.json",
            "-v" if verbose else "-q"
        ]
        
        try:
            result = subprocess.run(
                cmd,
                cwd=self.api_root,
                capture_output=True,
                text=True,
                env={**os.environ, **{
                    "SECRET_KEY": "test_secret_key_that_is_32_chars_long",
                    "JWT_SECRET_KEY": "test_jwt_secret_key_that_is_32_chars_long",
                    "DATABASE_URL": "sqlite:///test.db",
                    "DATABASE_URL_ASYNC": "sqlite+aiosqlite:///test.db",
                    "OPENAI_API_KEY": "test_key"
                }}
            )
            
            if result.returncode == 0:
                print("✅ Performance tests passed")
                self.test_results["suites"]["performance"] = {
                    "status": "passed",
                    "output": result.stdout
                }
                return True
            else:
                print("⚠️  Some performance benchmarks may have failed")
                print(result.stdout)
                self.test_results["suites"]["performance"] = {
                    "status": "warning",
                    "output": result.stdout,
                    "error": result.stderr
                }
                return True  # Don't fail CI for performance issues
                
        except Exception as e:
            print(f"⚠️  Error running performance tests: {e}")
            self.test_results["suites"]["performance"] = {
                "status": "error",
                "error": str(e)
            }
            return True  # Don't fail CI for performance issues
    
    def generate_coverage_report(self) -> None:
        """Generate comprehensive coverage report"""
        print("📊 Generating Coverage Report...")
        
        try:
            # Combine coverage files if they exist
            coverage_files = [
                "coverage_unit.json",
                "coverage_integration.json"
            ]
            
            total_coverage = 0
            coverage_count = 0
            
            for coverage_file in coverage_files:
                coverage_path = self.api_root / coverage_file
                if coverage_path.exists():
                    with open(coverage_path, 'r') as f:
                        coverage_data = json.load(f)
                        if 'totals' in coverage_data and 'percent_covered' in coverage_data['totals']:
                            total_coverage += coverage_data['totals']['percent_covered']
                            coverage_count += 1
            
            if coverage_count > 0:
                avg_coverage = total_coverage / coverage_count
                self.test_results["coverage"]["average"] = avg_coverage
                
                if avg_coverage >= 90:
                    print(f"✅ Excellent coverage: {avg_coverage:.1f}%")
                elif avg_coverage >= 80:
                    print(f"✅ Good coverage: {avg_coverage:.1f}%")
                elif avg_coverage >= 70:
                    print(f"⚠️  Moderate coverage: {avg_coverage:.1f}%")
                else:
                    print(f"❌ Low coverage: {avg_coverage:.1f}%")
                    
        except Exception as e:
            print(f"⚠️  Error generating coverage report: {e}")
    
    def generate_final_report(self) -> None:
        """Generate final test report"""
        print("\n" + "="*60)
        print("📋 COMPREHENSIVE TEST REPORT")
        print("="*60)
        
        # Count results
        passed = sum(1 for suite in self.test_results["suites"].values() if suite["status"] == "passed")
        failed = sum(1 for suite in self.test_results["suites"].values() if suite["status"] == "failed")
        warnings = sum(1 for suite in self.test_results["suites"].values() if suite["status"] == "warning")
        errors = sum(1 for suite in self.test_results["suites"].values() if suite["status"] == "error")
        
        self.test_results["summary"] = {
            "passed": passed,
            "failed": failed,
            "warnings": warnings,
            "errors": errors,
            "total": len(self.test_results["suites"])
        }
        
        # Print summary
        for suite_name, result in self.test_results["suites"].items():
            status_icon = {
                "passed": "✅",
                "failed": "❌",
                "warning": "⚠️ ",
                "error": "💥"
            }.get(result["status"], "❓")
            
            print(f"{status_icon} {suite_name.title()} Tests: {result['status'].upper()}")
        
        if "average" in self.test_results["coverage"]:
            print(f"📊 Average Coverage: {self.test_results['coverage']['average']:.1f}%")
        
        print(f"\n🎯 Summary: {passed} passed, {failed} failed, {warnings} warnings, {errors} errors")
        
        # Save detailed report
        report_file = self.api_root / "test_report.json"
        with open(report_file, 'w') as f:
            json.dump(self.test_results, f, indent=2)
        
        print(f"📁 Detailed report saved to: {report_file}")
        
        if failed > 0:
            print("\n❌ Some tests failed. Check the output above for details.")
            return False
        elif errors > 0:
            print("\n⚠️  Some test suites had errors. Check the output above.")
            return False
        else:
            print("\n✅ All tests passed successfully!")
            return True


def main():
    """Main test runner function"""
    parser = argparse.ArgumentParser(description="Run comprehensive file upload tests")
    parser.add_argument("--verbose", "-v", action="store_true", help="Verbose output")
    parser.add_argument("--unit-only", action="store_true", help="Run only unit tests")
    parser.add_argument("--integration-only", action="store_true", help="Run only integration tests")
    parser.add_argument("--security-only", action="store_true", help="Run only security tests")
    parser.add_argument("--performance-only", action="store_true", help="Run only performance tests")
    parser.add_argument("--skip-coverage", action="store_true", help="Skip coverage report")
    
    args = parser.parse_args()
    
    # Get project root (assuming script is in apps/api/tests/)
    script_dir = Path(__file__).parent
    project_root = script_dir.parent.parent.parent
    
    runner = TestRunner(str(project_root))
    
    print("🚀 Starting Comprehensive Test Suite for File Upload Functionality")
    print(f"📁 Project Root: {project_root}")
    print(f"⏰ Started at: {datetime.now().isoformat()}")
    print()
    
    success = True
    
    # Run selected test suites
    if args.unit_only:
        success = runner.run_unit_tests(args.verbose)
    elif args.integration_only:
        success = runner.run_integration_tests(args.verbose)
    elif args.security_only:
        success = runner.run_security_tests(args.verbose)
    elif args.performance_only:
        success = runner.run_performance_tests(args.verbose)
    else:
        # Run all test suites
        success &= runner.run_unit_tests(args.verbose)
        success &= runner.run_integration_tests(args.verbose)
        success &= runner.run_security_tests(args.verbose)
        success &= runner.run_performance_tests(args.verbose)
    
    # Generate coverage report
    if not args.skip_coverage:
        runner.generate_coverage_report()
    
    # Generate final report
    final_success = runner.generate_final_report()
    
    # Exit with appropriate code
    sys.exit(0 if (success and final_success) else 1)


if __name__ == "__main__":
    main()